src/daemon/daemon.cpp

```diff
@@ -2135,91 +2161,104 @@ void mp::Daemon::create_vm(const CreateRequest* request, grpc::ServerWriter<Crea
-                reply.set_create_message("Creating " + name);
+                reply.set_create_message("Preparing image for " + name);
                 server->Write(reply);
-                auto vm_image = config->vault->fetch_image(fetch_type, query, prepare_action, progress_monitor);
 
-                // TODO: make this come from the image host
-                const auto& driver = utils::get_driver_str();
-                mp::MemorySize disk_space;
-                if (driver == "lxd")
-                {
-                    disk_space = *checked_args.disk_space;
-                }
-                else
-                {
-                    disk_space = compute_final_image_size(vm_image, checked_args.disk_space);
-                }
+                return config->factory->prepare_source_image(source_image);
+            };
 
-                reply.set_create_message("Configuring " + name);
-                server->Write(reply);
+            auto fetch_type = config->factory->fetch_type();
 
-                // Generate a default network interface.
-                mp::NetworkInterface default_interface{"default", generate_unused_mac_address(), true};
+            CreateReply reply;
+            reply.set_create_message("Creating " + name);
+            server->Write(reply);
+            auto vm_image = config->vault->fetch_image(fetch_type, query, prepare_action, progress_monitor);
+
+            // TODO: make this come from the image host
+            const auto& driver = utils::get_driver_str();
+            mp::MemorySize disk_space;
+            if (driver == "lxd")
+            {
+                disk_space = *checked_args.disk_space;
+            }
+            else
+            {
+                disk_space = compute_final_image_size(vm_image, checked_args.disk_space);
+            }
 
-                // Generate MAC addresses for the interfaces on which it was not specified and put the modified
-                // interfaces in a new vector. Create another vector containing only the interfaces which must be
-                // configured via cloud-init.
-                std::vector<mp::NetworkInterface> extra_interfaces;
+            reply.set_create_message("Configuring " + name);
+            server->Write(reply);
 
-                for (auto iface : checked_args.extra_interfaces)
+            // To generate the network interfaces, we need to check first for repetition the MAC addresses
+            // which were specified by the user, and add them to the set of new MAC addresses. If it happens
+            // that we have repeated addresses, fail. Only after doing this we will be able to generate the
+            // rest of the addresses.
+            for (const auto& iface : checked_args.extra_interfaces)
+                if (!iface.mac_address.empty() &&
+                    allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end() &&
+                    !added_mac_addresses.insert(iface.mac_address).second)
```

Comment:
The second `&&` should be `||` instead. 

```suggestion
                if (!iface.mac_address.empty() &&
                    (allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end() ||
                    !added_mac_addresses.insert(iface.mac_address).second))
```
But please add such new code in new commits. Otherwise it's easier to miss in review, especially because this many changes is also messing with diff anchors, so the diffs include a lot of code that wasn't touched. Separating changes in their own commits helps with that and would allow me to quickly isolate new code since I last commented.